File: clients/kotlin/ares-client/src/main/kotlin/ares/client/ByteUtil.kt
package ares.client

object ByteUtil {
    fun u32be(n: Int): ByteArray =
        byteArrayOf((n ushr 24).toByte(), (n ushr 16).toByte(), (n ushr 8).toByte(), n.toByte())

    fun lp(b: ByteArray): ByteArray = u32be(b.size) + b

    fun hex(b: ByteArray): String {
        val sb = StringBuilder(b.size * 2)
        for (x in b) sb.append("%02x".format(x.toInt() and 0xff))
        return sb.toString()
    }

    fun fromHex(s: String): ByteArray? {
        if (s.length % 2 != 0) return null
        val out = ByteArray(s.length / 2)
        var i = 0
        while (i < s.length) {
            val v = s.substring(i, i + 2).toIntOrNull(16) ?: return null
            out[i / 2] = v.toByte(); i += 2
        }
        return out
    }
}

File: clients/swift/Sources/AresClient/ByteUtil.swift
import Foundation

/// Byte helpers matching the ARES wire encoding (big-endian length prefixes, hex).
public enum ByteUtil {
    /// 4-byte big-endian encoding of n.
    public static func u32be(_ n: UInt32) -> Data {
        Data([UInt8((n >> 24) & 0xff), UInt8((n >> 16) & 0xff),
              UInt8((n >> 8) & 0xff), UInt8(n & 0xff)])
    }

    /// Length-prefix: u32be(len) || b.
    public static func lp(_ b: Data) -> Data {
        var out = u32be(UInt32(b.count))
        out.append(b)
        return out
    }

    /// Lowercase hex string of d.
    public static func hex(_ d: Data) -> String {
        d.map { String(format: "%02x", $0) }.joined()
    }

    /// Decode a hex string to Data; nil on malformed input.
    public static func fromHex(_ s: String) -> Data? {
        guard s.count % 2 == 0 else { return nil }
        var out = Data(capacity: s.count / 2)
        var idx = s.startIndex
        while idx < s.endIndex {
            let next = s.index(idx, offsetBy: 2)
            guard let byte = UInt8(s[idx..<next], radix: 16) else { return nil }
            out.append(byte)
            idx = next
        }
        return out
    }
}
